rasa/nlu/featurizers/sparse_featurizer/lexical_syntactic_featurizer.py

```diff
@@ -69,241 +88,424 @@ def required_components(cls) -> List[Type[Component]]:
+        sentence_features = []
 
-        # build vocabulary of features
-        feature_vocabulary = self._build_feature_vocabulary(all_features)
+        # in case of an even number we will look at one more word before,
+        # e.g. window size 4 will result in a window range of
+        # [-2, -1, 0, 1] (0 = current word in sentence)
+        window_size = len(self._feature_config)
+        half_window_size = window_size // 2
+        window_range = range(-half_window_size, half_window_size + window_size % 2)
+        assert len(window_range) == window_size
 
-        # assign a unique index to each feature value
-        return self._map_features_to_indices(feature_vocabulary)
+        for anchor in range(len(tokens)):
+
+            token_features: Dict[Tuple[int, Text], Text] = {}
+
+            for window_position, relative_position in enumerate(window_range):
+                absolute_position = anchor + relative_position
+
+                # skip, if current_idx is pointing to a non-existing token
+                if absolute_position < 0 or absolute_position >= len(tokens):
+                    continue
+
+                token = tokens[absolute_position]
+                for feature_name in self._feature_config[window_position]:
+                    token_features[
+                        (window_position, feature_name)
+                    ] = self._extract_raw_features_from_token(
+                        token=token,
+                        feature_name=feature_name,
+                        token_position=absolute_position,
+                        num_tokens=len(tokens),
+                    )
+
+            sentence_features.append(token_features)
+
+        return sentence_features
 
     @staticmethod
-    def _map_features_to_indices(
-        feature_vocabulary: Dict[Text, List[Text]]
-    ) -> Dict[Text, Dict[Text, int]]:
-        feature_to_idx_dict = {}
+    def _build_feature_to_index_map(
+        feature_vocabulary: Dict[Tuple[int, Text], Set[Text]]
+    ) -> Dict[Tuple[int, Text], Dict[Text, int]]:
+        """Creates a nested dictionary for mapping raw features to indices.
+
+        Args:
+          feature_vocabulary: a mapping from tuples of positions (in the window) and
+            supported feature names to the set of possible feature values
+        Returns:
+           a nested mapping that maps from tuples of positions (in the window) and
+           supported feature names to "raw feature to index" mappings, i.e.
+           mappings that map the respective raw feature values to unique indices
+           (where `unique` means unique with respect to all indices in the
+           *nested* mapping)
+        """
+        # Note that this will only sort the top level keys - and we keep
+        # doing it to ensure consistently with what was done before)
+        ordered_feature_vocabulary: OrderedDict[
+            Tuple[int, Text], Set[Text]
+        ] = OrderedDict(sorted(feature_vocabulary.items()))
```

Comment:
pycharm actually still complains but it's a false one and should be clear from the type definitions 🤦🏻